src/qt3utils/datagenerators/princeton.py

```diff
@@ -0,0 +1,344 @@
+        self._automation.Dispose()
+        logger.info('Closed AddInProcess.exe')
+
+    def set(self, setting: str, value: Any) -> None:
+        """
+        Helper function for setting experiment parameters with basic value
+        checking.
+        """
+        if self.experiment.Exists(setting):
+            if self.experiment.IsValid(setting, value):
+                self.experiment.SetValue(setting, value)
+            else:
+                # TODO: might need to add a proper exception
+                logger.error(f'Invalid value: {value} for setting: {setting}.')
+        else:
+            # TODO: might need to add a proper exception
+            logger.error(f'Invalid setting: {setting}.')
+
+    def get(self, setting: Any) -> Any:
+        """
+        Helper function for getting experiment parameters with a basic check
+        to see if a specific setting is valid.
+        """
+        if self.experiment.Exists(setting):
+            value = self.experiment.GetValue(setting)
+        else:
+            value = []
+            logger.error(f'Invalid setting: {setting}.')
+        return value
+
+    def load_experiment(self, value: str) -> None:
+        self.experiment.Load(value)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationAttachDate, False)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationAttachTime, False)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationAttachIncrement, True)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationIncrementNumber, 0)
+
+    def file_setup(self) -> None:
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationBaseFileName, str(uuid.uuid4()))
+    
+    #TODO: Need to actually implement the three methods below so that you can:
+    #      - save a new experiment
+    #      - save what you are currently working on
+    #      - stop the scan 
+        
+    def save_current_experiment(self) -> None:
+        self.experiment.Save()
+    
+    def save_new_experient(self, value) -> None:
+        self.experiment.SaveAs(value)
+
+    def stop_scan(self) -> None:
+        self.experiment.Stop()
+
+    #NOTE: Thw while loop in here will not be needed once you fix the "FileNameGeneration" problem.
+    def start_acquisition_and_wait(self) -> None:
+        """
+        Starts the acquisition process and waits until it is completed before carrying on.
+        """
+        acquisition_time = self.get(lf.AddIns.CameraSettings.ShutterTimingExposureTime)
+        num_frames = self.get(lf.AddIns.ExperimentSettings.AcquisitionFramesToStore)
+        self.experiment.Acquire()
+        
+        sleep(0.001 * acquisition_time * num_frames)  # waiting for the exposure duration
+        
+        while self.experiment.IsRunning:
+            sleep(0.1)  # checking if the experiment is still running
+
+    def process_acquired_data(self) -> Union[np.ndarray, Any]:
```

Comment:
I don't understand this type hint. Should it just be np.ndarray?

Reply:
Yes, will fix